client/protocols/hinting.py

```diff
@@ -0,0 +1,185 @@
+
+            if not hint_info['elgible']:
+                continue
+            log.info('Prompting for hint on %s', question)
+
+            if confirm("Check for hints on {}?".format(question)):
+                hint_info['accept'] = True
+                print("Thinking... (could take upto 15 seconds)", end='')
+                try:
+                    response = self.query_server(messages, question)
+                    hint_info['response'] = response
+
+                    hint = response['message']
+                    pre_prompt = response['pre-prompt']
+                    post_prompt = response['post-prompt']
+                    log.info("Hint server response: {}".format(response))
+                    if not hint and not pre_prompt:
+                        print("\nSorry. No hints found for {}".format(question))
+                        continue
+
+                    if pre_prompt:
+                        print("\r-- While we wait, respond to this question."
+                              " When you are done typing, press Enter.")
+                        if not prompt_user(pre_prompt, hint_info):
+                            # Do not provide hint, if no response from user
+                            continue
+
+                    # Provide padding for the the hint
+                    print("\n{}\n".format(hint))
+
+                    if post_prompt:
+                        prompt_user(post_prompt, hint_info)
+
+                except urllib.error.URLError:
+                    log.debug("Network error while fetching hint")
+                    hint_info['fetch_error'] = True
+                    print("\r\nCould not get a hint.")
+            else:
+                log.info('Declined Hints for %s', question)
+                hint_info['accept'] = False
+
+    def query_server(self, messages, test):
+        access_token, _, _ = auth.get_storage()
+        user = auth.get_student_email(access_token) or access_token
+        if user:
+            user = hash(user)
+        data = {
+            'assignment': self.assignment.endpoint,
+            'test': test,
+            'messages': messages,
+            'user': user
+        }
+
+        serialized_data = json.dumps(data).encode(encoding='utf-8')
+
+        address = self.HINT_SERVER + self.HINT_ENDPOINT
+
+        log.info('Sending hint request to %s', address)
+        request = urllib.request.Request(address)
+        request.add_header("Content-Type", "application/json")
+
+        response = urllib.request.urlopen(request, serialized_data, 10)
+        return json.loads(response.read().decode('utf-8'))
+
+def prompt_user(query, results):
+    try:
+        response = None
+        short_resps = 0
+        while not response or response == '':
+            response = input("{}\nYour Response: ".format(query))
+            if not response or response == '':
+                short_resps += 1
+                if short_resps > 2:
```

Comment:
Explain why you are breaking here (e.g. "don't ask more than 2 times, to avoid annoying the user")